pkg/build/strategies/sti/sti.go

```diff
@@ -568,77 +554,43 @@ func (builder *STI) Execute(command string, user string, config *api.Config) err
 			return nil
 		}
 	} else {
-		close(injectionComplete)
+		close(injectionError)
 	}
 
-	wg := sync.WaitGroup{}
 	if !config.LayeredBuild {
-		wg.Add(1)
-		uploadDir := filepath.Join(config.WorkingDir, "upload")
-		// TODO: be able to pass a stream directly to the Docker build to avoid the double temp hit
 		r, w := io.Pipe()
+		opts.Stdin = r
+
 		go func() {
-			// reminder, multiple defers follow a stack, LIFO order of processing
-			defer wg.Done()
 			// Wait for the injections to complete and check the error. Do not start
 			// streaming the sources when the injection failed.
-			<-injectionComplete
-			if injectionError != nil {
+			if <-injectionError != nil {
+				w.Close()
 				return
 			}
 			glog.V(2).Info("starting the source uploading ...")
-			var err error
-			defer func() {
-				w.CloseWithError(err)
-				if r := recover(); r != nil {
-					glog.Errorf("recovered panic: %#v", r)
-				}
-			}()
-			err = builder.tar.CreateTarStream(uploadDir, false, w)
+			uploadDir := filepath.Join(config.WorkingDir, "upload")
+			w.CloseWithError(builder.tar.CreateTarStream(uploadDir, false, w))
 		}()
-
-		opts.Stdin = r
 	}
 
-	go func(reader io.Reader) {
-		scanner := bufio.NewReader(reader)
-		// Precede build output with newline
-		glog.Info()
-		for {
-			text, err := scanner.ReadString('\n')
-			if err != nil {
-				// we're ignoring ErrClosedPipe, as this is information
-				// the docker container ended streaming logs
-				if glog.Is(2) && err != io.ErrClosedPipe && err != io.EOF {
-					glog.Errorf("Error reading docker stdout, %#v", err)
-				}
-				break
-			}
-			// Nothing is printed when the quiet option is set
-			if config.Quiet {
-				continue
-			}
-			glog.Info(strings.TrimSpace(text))
+	dockerpkg.StreamContainerIO(outReader, nil, func(s string) {
+		if !config.Quiet {
+			glog.Info(strings.TrimSpace(s))
 		}
-		// Terminate build output with new line
-		glog.Info()
-
-	}(outReader)
+	})
 
-	go dockerpkg.StreamContainerIO(errReader, &errOutput, func(a ...interface{}) { glog.Info(a...) })
+	c := dockerpkg.StreamContainerIO(errReader, &errOutput, func(s string) { glog.Info(s) })
 
 	err := builder.docker.RunContainer(opts)
+	outWriter.Close()
+	errWriter.Close()
 	if e, ok := err.(errors.ContainerError); ok {
-		// even with deferred close above, close errReader now so we avoid data race condition on errOutput;
-		// closing will cause StreamContainerIO to exit, thus releasing the writer in the equation
-		errReader.Close()
-		return errors.NewContainerError(config.BuilderImage, e.ErrorCode, errOutput)
-	}
-	// Do not wait for source input if the container times out.
-	// FIXME: this potentially leaks a goroutine.
-	if !util.IsTimeoutError(err) {
-		wg.Wait()
+		// Must wait for StreamContainerIO goroutine above to exit before reading errOutput.
+		<-c
+		err = errors.NewContainerError(config.BuilderImage, e.ErrorCode, errOutput)
```

Comment:
@rhcarvalho thx, @jim-minter just fyi ... I made a change in this area as well yesterday ... https://github.com/openshift/source-to-image/pull/631 
It looks like your change fixes things as well :) 👍